pkg/config/config.go

```diff
@@ -1,26 +1,75 @@
 package config
 
+import (
+	"fmt"
+
+	"github.com/caarlos0/env"
+	"github.com/golang/glog"
+)
+
 type DaemonConfig struct {
-	Plugin         string `json:"plugin"`          // Subnet manager plugin name
-	PeriodicUpdate int    `json:"periodic_update"` // Interval between every check for the added and deleted pods
-	GuidPool       GuidPoolConfig
-	SubnetManager  SubnetManagerPluginConfig
+	Plugin         string                    `json:"plugin" env:"PLUGIN"`                   // Subnet manager plugin name
+	PeriodicUpdate int                       `json:"periodic_update" env:"PERIODIC_UPDATE"` // Interval between every check for the added and deleted pods
+	GuidPool       GuidPoolConfig            `json:"guid_pool"`
+	SubnetManager  SubnetManagerPluginConfig `json:"subnet_manager"`
 }
 
 type GuidPoolConfig struct {
-	GuidRangeStart string `json:"guid_range_start"` // First guid of the pool
-	GuidRangeEnd   string `json:"guid_range_end"`   // Last of the guid pool
+	GuidRangeStart string `json:"guid_range_start" env:"GUID_RANGE_START"` // First guid of the pool
+	GuidRangeEnd   string `json:"guid_range_end" env:"GUID_RANGE_END"`     // Last of the guid pool
 }
 
 type SubnetManagerPluginConfig struct {
 	UFMConfig
 }
 
 type UFMConfig struct {
-	Username    string `json:"ufm_username"`    // Username of ufm
-	Password    string `json:"ufm_password"`    // Password of ufm
-	Address     string `json:"ufm_address"`     // IP address or hostname of ufm servere
-	Port        int    `json:"ufm_port"`        // REST API port of ufm
-	HttpSchema  string `json:"ufm_httpSchema"`  // http or https
-	Certificate string `json:"ufm_certificate"` // Certificate of ufm
+	Username    string `json:"ufm_username" env:"UFM_USERNAME"`       // Username of ufm
+	Password    string `json:"ufm_password" env:"UFM_PASSWORD"`       // Password of ufm
+	Address     string `json:"ufm_address" env:"UFM_ADDRESS"`         // IP address or hostname of ufm server
+	Port        int    `json:"ufm_port" env:"UFM_PORT"`               // REST API port of ufm
+	HttpSchema  string `json:"ufm_http_Schema" env:"UFM_HTTP_SCHEMA"` // http or https
+	Certificate string `json:"ufm_certificate" env:"UFM_CERTIFICATE"` // Certificate of ufm
+}
+
+var supportedPlugins = []string{"noop", "ufm"}
+
+func (dc *DaemonConfig) ReadConfig() error {
+	glog.Info("ReadConfig():")
+	return env.Parse(dc)
+}
+
+func (dc *DaemonConfig) ValidateConfig() error {
+	glog.Info("ValidateConfig():")
+	if dc.PeriodicUpdate <= 0 {
```

Comment:
do we want to have a default value for it ?

Reply:
added

Reply:
Where you define defualt value